### PR TITLE
Removed the need for a .env file

### PR DESCRIPTION
**Pull request by mdriessen**
The .env file and the Flex recipe aren't needed anymore after these changes, but let me try to explain why I've done them this way.

The parameters.yml / .env file had two purposes:
1) Provide easy and fast access to the version number; we don't want to call the providers every time we need the version number
2) Provide a persistent storage for the version number when not using Git or Revision files (hence the parameterProvider we had)

The version bump command also has two purposes:
1) Update the parameter.yml file with the current version fetched from other providers
2) Manually bump the version number with the major/minor/patch options

Both 1) points are easily solved using the filesystem cache that Symfony provides. If the cache is empty the version number is fetched from the providers and if it's not empty we have the fast access we want. We also don't need the bump command anymore because the cache is cleared after every deployment so the version gets updated automatically.

However using a cache for points 2) doens't make sense, it's not persistent. I think the most elegant way to solve this is by using a VERSION file. I thought about just using the REVISION file instead, but since Capistrano uses that file I think it's better to keep it seperated. Also, I wanted this VersionProvider at the highest priority for manually bumping the version, while the RevisionProvider is below the GitDescribe provider.

Long story short... when using Git tags or the REVISION file to provide the version, everything should just work out of the box. The bump command is only there for manually bumping the version, it creates a VERSION file which can be committed and deployed alongside the code. I also created a status command to check the version in the cache.

@shivas If you approve these changes, I'll remove the Flex recipe before you merge things. We don't want to run the bump command anymore and the recipe automates that part.

### CODE REVIEW

**Comment by shivas** [comment]
@mdriessen yeah, looks good to me, sorry really busy RL

File: Command/VersionBumpCommand.php [hunk 1]
```diff
@@ -44,83 +44,85 @@ protected function configure()
...
+            ->addOption('dry-run', 'd', InputOption::VALUE_NONE, 'Dry run, does not update VERSION file')
             ->addOption('major', null, InputOption::VALUE_OPTIONAL, 'Bump MAJOR version by given number', 0)
             ->addOption('minor', null, InputOption::VALUE_OPTIONAL, 'Bump MINOR version by given number', 0)
             ->addOption('patch', null, InputOption::VALUE_OPTIONAL, 'Bump PATCH version by given number', 0)
             ->addOption('prerelease', null, InputOption::VALUE_IS_ARRAY | InputOption::VALUE_OPTIONAL, 'Add PRERELEASE to version', array())
             ->addOption('build', null, InputOption::VALUE_IS_ARRAY | InputOption::VALUE_OPTIONAL, 'Add BUILD to version', array());
     }
 
+    /**
+     * Manually bump application version
```
Comment (stof) [inline Command/VersionBumpCommand.php]: this does not match the command description. Is it expected ?

File: Resources/config/services.xml [hunk 2]
```diff
@@ -5,26 +5,44 @@
            xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
 
     <services>
+        <service id="Shivas\VersioningBundle\Formatter\FormatterInterface"
+                 alias="Shivas\VersioningBundle\Formatter\GitDescribeFormatter"/>
+        <service id="Shivas\VersioningBundle\Writer\WriterInterface"
+                 alias="Shivas\VersioningBundle\Writer\VersionWriter"/>
```
Comment (stof) [inline Resources/config/services.xml]: these should be marked as private explicitly, to be private even on 3.4 (rather than half private for the BC layer)

File: Resources/config/services.xml [hunk 3]
```diff
@@ -5,26 +5,44 @@
...
-            <tag name="shivas_versioning.provider" alias="revision" priority="-25"/>
+            <tag name="shivas_versioning.provider" alias="revision" priority="-50"/>
         </service>
 
         <service id="Shivas\VersioningBundle\Provider\InitialVersionProvider">
-            <tag name="shivas_versioning.provider" alias="init" priority="-50"/>
+            <tag name="shivas_versioning.provider" alias="init" priority="-75"/>
+        </service>
+
+        <service id="Shivas\VersioningBundle\Command\StatusCommand">
```
Comment (stof) [inline Resources/config/services.xml]: For shared bundles, it is actually recommended to avoid using FQCN as service ids, except for types that you actually want to expose to autowiring of projects (so probably only FormatterInterface, WriterInterface and VersionManager). This allows keeping `debug:autowiring` clean for projects (not exposing all your commands in it for instance)

File: Resources/config/services.xml [hunk 4]
```diff
@@ -5,26 +5,44 @@
            xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
 
     <services>
+        <service id="Shivas\VersioningBundle\Formatter\FormatterInterface"
+                 alias="Shivas\VersioningBundle\Formatter\GitDescribeFormatter"/>
+        <service id="Shivas\VersioningBundle\Writer\WriterInterface"
+                 alias="Shivas\VersioningBundle\Writer\VersionWriter"/>
+
         <service id="Shivas\VersioningBundle\Service\VersionManager">
+            <argument type="service" id="Symfony\Component\Cache\Adapter\AdapterInterface"/>
```
Comment (stof) [inline Resources/config/services.xml]: you should not get the main cache pool (which is the `cache.app`, not cleaned on deployment btw), but create your own cache pool (so that keys in it are isolated from the cache keys of the project to avoid conflicts). And this cache pool should be based on `cache.system` if you want it to be cleared automatically on deployment.
This is very easy: https://github.com/symfony/symfony/blob/v4.0.7/src/Symfony/Bundle/FrameworkBundle/Resources/config/cache.xml#L18-L20

File: Resources/config/services.xml [hunk 5]
```diff
@@ -1,61 +1,63 @@
...
+        <service id="shivas_versioning.provider.init" class="Shivas\VersioningBundle\Provider\InitialVersionProvider" public="false">
             <tag name="shivas_versioning.provider" alias="init" priority="-75"/>
         </service>
 
-        <service id="Shivas\VersioningBundle\Command\StatusCommand">
+        <service id="shivas_versioning.cache.version" parent="cache.system" public="false">
+            <tag name="cache.pool" />
+        </service>
+
+        <service id="shivas_versioning.command.status" class="Shivas\VersioningBundle\Command\StatusCommand" public="false">
```
Comment (stof) [inline Resources/config/services.xml]: you could use file-level defaults to mark all services as private, as the bundle requires 3.4+ anyway

File: Resources/config/services.xml [hunk 6]
```diff
@@ -1,61 +1,63 @@
...
+
+        <service id="shivas_versioning.formatter.git" class="Shivas\VersioningBundle\Formatter\GitDescribeFormatter" public="false"/>
+        <service id="shivas_versioning.writer.version" class="Shivas\VersioningBundle\Writer\VersionWriter" public="false">
             <argument>%kernel.project_dir%</argument>
         </service>
 
-        <service id="Shivas\VersioningBundle\Provider\VersionProvider">
+        <service id="shivas_versioning.provider.version" class="Shivas\VersioningBundle\Provider\VersionProvider" public="false">
             <argument>%kernel.project_dir%</argument>
             <tag name="shivas_versioning.provider" alias="version" priority="0"/>
```
Comment (stof) [inline Resources/config/services.xml]: I think this provider should have a higher priority, to stay before providers registered without a priority, as it is meant to win over anything else to make manual versioning working.

File: Service/VersionManager.php [hunk 7]
```diff
@@ -123,33 +124,43 @@ public function getActiveProvider()
...
+        try {
+            $cacheItem = $this->cache->getItem('version');
+            $cacheItem->set($version);
 
-        $this->cache->save($cacheItem);
-        $this->writer->write($version);
+            $this->cache->save($cacheItem);
+            $this->writer->write($version);
+        } catch (InvalidArgumentException $e) {
+            throw new RuntimeException('Could not write version cache item: ' . $e->getMessage());
```
Comment (stof) [inline Service/VersionManager.php]: I would keep the original exception as previous exception to make debugging easier.

File: Resources/config/services.xml (hunk 6 above)
Comment (mdriessen) [reply Resources/config/services.xml]: Good catch, that was indeed the intention! Thanks for the useful feedback!

**Comment by shivas** [comment]
@stof thanks for your input, you think it's looking ok now after @mdriessen commits?
